runtime/src/main/java/org/corfudb/runtime/collections/Table.java

```diff
@@ -351,17 +350,20 @@ public int count() {
     @Nonnull
     List<CorfuStoreEntry<K, V, M>> scanAndFilterByEntry(
             @Nonnull final Predicate<CorfuStoreEntry<K, V, M>> entryPredicate) {
-        return corfuTable.scanAndFilterByEntry(recordEntry ->
-                entryPredicate.test(new CorfuStoreEntry<>(
-                        recordEntry.getKey(),
-                        recordEntry.getValue().getPayload(),
-                        recordEntry.getValue().getMetadata())))
-                .parallelStream()
-                .map(entry -> new CorfuStoreEntry<>(
-                        entry.getKey(),
-                        entry.getValue().getPayload(),
-                        entry.getValue().getMetadata()))
-                .collect(Collectors.toList());
+        try(Stream<Map.Entry<K, CorfuRecord<V, M>>> stream = corfuTable.entryStream()) {
+            return CorfuTable.pool.submit(() -> stream
+                    .filter(recordEntry ->
+                            entryPredicate.test(new CorfuStoreEntry<>(
+                                    recordEntry.getKey(),
+                                    recordEntry.getValue().getPayload(),
+                                    recordEntry.getValue().getMetadata())))
+                    .parallel()
```

Comment:
Question: why not put parallel() in front of filter()?

Reply:
I think it automatically detects it. Try this

```java

    public final static ForkJoinPool pool = new ForkJoinPool(Math.max(Runtime.getRuntime().availableProcessors() - 1, 1),
            pool -> {
                final ForkJoinWorkerThread worker = ForkJoinPool.defaultForkJoinWorkerThreadFactory.newThread(pool);
                worker.setName("CorfuTable-Forkjoin-pool-" + worker.getPoolIndex());
                return worker;
            }, null, true);

    @Test
    public void blah() {
        StreamingMapDecorator map = new StreamingMapDecorator<>(new HashMap<>());

        int numKeys = 100;
        for (int x = 0; x < numKeys; x++) {
            map.put(x + "", x + "");
        }

        Object res = pool.submit(() -> map.entryStream()
                .filter(t -> {
                    System.out.println(Thread.currentThread().getId() + " filter " + t);
                    return true;
                })
                .map(t -> {
                    System.out.println(Thread.currentThread().getId() + " map " + t);
                    return t;})
                .collect(Collectors.toList())).join();

    }
```

Although I'm still not certain on how this detection works.

Reply:
Right. I ran the test and found that no matter where I put parallel(), the other intermediate operations (filter and map) are always parallelized. Amazing.